jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/DefaultServlet.java

```diff
@@ -487,44 +473,70 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
                 }
 
                 // no content
-                resp.sendError(404);
+                httpServletResponse.sendError(404);
             }
             else
             {
-                ServletCoreRequest coreRequest = new ServletCoreRequest(req);
-                ServletCoreResponse coreResponse = new ServletCoreResponse(coreRequest, resp, included);
-
+                // lookup the core request and response as wrapped by the ServletContextHandler
+                ServletContextRequest servletContextRequest = ServletContextRequest.getServletContextRequest(httpServletRequest);
+                ServletContextResponse servletContextResponse = servletContextRequest.getServletContextResponse();
+                ServletChannel servletChannel = servletContextRequest.getServletChannel();
+
+                // If the servlet request has not been wrapped,
+                // we can use the core request directly,
+                // otherwise wrap the servlet request as a core request
+                Request coreRequest = httpServletRequest instanceof ServletApiRequest
+                    ? servletChannel.getRequest()
+                    : new ServletCoreRequest(httpServletRequest);
+
+                // If the servlet response has not been wrapped and has not been written to,
+                // we can use the core response directly,
+                // otherwise wrap the servlet response as a core response
+                Response coreResponse = httpServletResponse instanceof ServletApiResponse && !servletContextResponse.isWritingOrStreaming()
```

Comment:
Since the `httpServletResponse instanceof ServletApiResponse && !servletContextResponse.isWritingOrStreaming()` logic is re-used a few lines below as `coreResponse instanceof ServletCoreResponse` maybe it would be clearer to introduce a boolean variable named `isResponseWrapped`?